Feature/m0081 meja 2407 managedc doc update v0.2.0 (#823)
MicroEJ/docs: ApplicationDeveloperGuide/managedc/communication.rst

.. _managedc.communication:

Communication Between Java and Managed C
========================================

The MicroEJ Core Engine allows to communicate between Java and Managed C. Java static methods can be used from Managed C code and vice-versa.

.. _managedc.communication.java_to_managedc:

Calling Java from Managed C
---------------------------

The MicroEJ Core Engine allows to expose Java static methods to Managed C. Functions has to be declared as ``extern`` in the Managed C 
source code. Here is a C source code example:

.. code:: c

   /* Extern functions implemented in Java -----*/
   extern void delay(int ms);

> Link between Managed C function declaration and Java static methods can be done statically using the following C macro definition:

Author: Here is my proposal
```suggestion
Link between Managed C function declaration and Java static methods can be done explicitly using the following C macro definition:
```

.. code:: c

   #define __IMPORT(class, method) \
    __attribute__((__import_module__(#class), __import_name__(#method)))

* ``class``: 

  * Java `fully qualified name <https://docs.oracle.com/javase/specs/jls/se11/html/jls-6.html#jls-6.7>`__ of the class containing the Java static method (e.g: ``com.mycompany.MyApp``).
  * An empty string (``""``) or ``"env"``. In that case, Java static method is searched in the current annotated ``@ManagedCModule`` Java class.

> * ``method``: Java static method name. Empty string XXX??XXX

Author: ```suggestion
* ``method``: Java static method name.
```

      +------------------------+-----------------+
      | float (32 bit)         | float (32 bit)  |
      +------------------------+-----------------+
      | double (64 bit)        | double (64 bit) |
      +------------------------+-----------------+

   A SOAR error will be triggered in case of Managed C function parameter(s) and return types do not matched excatly the same Java method parameter(s) and return types.  

Here is an example showing how to write Java and C source code when C macro parameters ``class`` and ``method`` are specified:

- Java source code (``MyApp.java``):

   .. code:: java

      package com.mycompany;

      public class MyApp {

         public static void delay(int ms) {
            try {
               Thread.sleep(ms);
            } catch (InterruptedException e) {
               e.printStackTrace();
            }
         }

      }

- C source code (``my_app.c``):

   .. code:: c

      #define __IMPORT(class, method) \
      __attribute__((__import_module__(class), __import_name__(method)))

      /* Extern functions implemented in Java -----*/
      extern void delay(int ms) __IMPORT("com.mycompany.MyApp" ,"delay");

.. note:: 

>    Java annotation ``@ManagedCModule("my_app.mc")`` can be put on the Java class but is useless for this example.

Author: ```suggestion
   Java annotation ``@ManagedCModule("my_app.mc")`` should be put some where, in this class or in another class.
```

Here is an example showing how to write Java and C source code when the C macro parameter ``class`` is an empty string:

- Java source code (``MyApp.java``): same as previous example

   .. code:: java

      package com.mycompany;

      @ManagedCModule("my_app.mc")
      public class MyApp {

         public static void delay(int ms) {
            try {
               Thread.sleep(ms);
            } catch (InterruptedException e) {
               e.printStackTrace();
            }
         }

      }

- C source code (``my_app.c``):

   .. code:: c

      #define __IMPORT(class, method) \
      __attribute__((__import_module__(class), __import_name__(method)))

      /* Extern functions implemented in Java -----*/
      extern void delay(int ms) __IMPORT("" ,"delay");

> Link between Managed C function declaration and Java static methods can also be done dynamically using ``-Wl,--allow-undefined`` 

Author: ```suggestion
Link between Managed C function declaration and Java static methods can also be done implicitly using ``-Wl,--allow-undefined`` 
```

C compiler option (see :ref:`managedc.compilation.command_line_options` ). No need to declare and use ``__IMPORT(class, method)`` C macro 

> in that case. Java type resolution will look at the ``@ManagedCModule`` declarations in Java source code to find the Java class 

Author: ```suggestion
in that case. Java static method is searched in the annotated ``@ManagedCModule`` Java class on the Managed C module.
```

> in that case. Java type resolution will look at the ``@ManagedCModule`` declarations in Java source code to find the Java class 

Reviewer: What about something like
```
in that case a Java static method is searched in the Java class  which refers to the current Managed C module.
```

Here is an example:

- C source code (``my_app.c``):

   .. code:: c

      int add(int a, int b) {
         return a + b;
      }

- Java source code (``MyApp.java``):

   .. code:: java

      package com.mycompany;

      import com.microej.managedc.ManagedCFunction;
      import com.microej.managedc.ManagedCModule;

      @ManagedCModule("my_app.mc")
      public class MyApp {

         public static void main(String[] args) {
            int a = 1;
            int b = 2;
            // Call and return result of the "add" Managed C function
            System.out.println(a + " + " + b + " = "+ add(a, b));
         }

         @ManagedCFunction
         public static native int add(int a, int b);

      }

.. note:: 

   A SOAR error will be triggered if no Java method or Managed C function found.

You should see the following output when launching the Java application:

   .. code:: console

          1 + 2 = 3

.. _managedc.communication.managedc_memory:

Manipulate Managed C Memory from Java
-------------------------------------

The MicroEJ Core Engine allows to expose Managed C memory to Java. A Managed C module contains 

> one (and only one) memory. This Managed C module memory is automatically generated by the C compiler 

Author: ```suggestion
at most one memory. This Managed C module memory is automatically generated by the C compiler 
```

according to C source code and C compiler options. On Java side, Managed C module memory can be seen by
using ``@ManagedCMemory`` annotation on a Java static byte array field declaration. 

.. note:: 

>    A SOAR error will occurred if ``@ManagedCMemory`` is not strictly followed by a Java static byte array declaration.

Reviewer: what happen when we have the annotation and the managedC module does not have a memory

Author: C compiler always generate a memory section.
